docs/proposals/docker-registry-2-4.md

```diff
@@ -0,0 +1,57 @@
+# Support for Docker Registry 2.4
+## Abstract
+
+A proposal to update the docker/distribution to v2.4.0.
+
+## Motivation
+
+Primary goal is to migrate to Schema 2 [1].
+
+## Compatibility
+
+docker/distribution maintains compatibility with old clients that don't know
+how to work with Schema 2.
+
+##### Docker 1.9
+
+When the manifest is pulled by digest or tag with any docker version, a Schema 1 manifest will be returned.
+
+##### Docker 1.10
+
+Docker Engine 1.10 tries to construct a manifest in the new format when pushing an image.
+If uploading this manifest fails, presumably because the registry only supports the old format,
+Docker will fall back to uploading a manifest in the old format.
+
+When the manifest is pulled by digest or tag with Docker Engine 1.10, a Schema 2 manifest will be returned.
+
+When the manifest is pulled by tag with Docker Engine 1.9 and older, the manifest is converted on-the-fly
+to Schema 1 and sent with the response. Docker 1.9 is compatible with this older format.
+
+If a manifest is pulled from a registry by digest with Docker Engine 1.9 and older,
+and the manifest was pushed with Docker Engine 1.10, a security check will cause the Engine
+to receive a manifest it cannot use and the pull will fail.
+
+##### Openshift registry
+
+For users who do not want to use Schema 2 at all: we can reject Schema 2 by config parameter and
+Docker Engine 1.10 will push the images in old format in accordance with the push strategy.
+
+## API changes
+
+* `MediaType` needs to be added to `ImageLayer`.
+* `Image.DockerImageMetadata` will be empty (except ID and Size). We don't have any data to fill it.
```

Comment:
Why do we need another query?  We can just inline it in the image struct.

Let me put it this way:

* The public OpenShift API needs to continue to efficiently return current
image metadata via the Image call - so all the relevant data needs to be in
the same etcd struct (manifest, sub data, etc)
* The registry should not store data for images in multiple places (same
etcd doc) and should not be using the filesystem for links or other
constructs that can be described by the image stored in etcd
* We have to be able to convert our etcd image struct into either schema1
or 2 on demand.
* We have to be able to read old data in schema1 (old images) and then
seamlessly return schema2


On Thu, May 5, 2016 at 12:04 PM, Alexey Gladkov <notifications@github.com>
wrote:

> In docs/proposals/docker-registry-2-4.md
> <https://github.com/openshift/origin/pull/8751#discussion_r62209279>:
>
> > +When the manifest is pulled by tag with Docker Engine 1.9 and older, the manifest is converted on-the-fly
> > +to Schema 1 and sent with the response. Docker 1.9 is compatible with this older format.
> > +
> > +If a manifest is pulled from a registry by digest with Docker Engine 1.9 and older,
> > +and the manifest was pushed with Docker Engine 1.10, a security check will cause the Engine
> > +to receive a manifest it cannot use and the pull will fail.
> > +
> > +##### Openshift registry
> > +
> > +For users who do not want to use Schema 2 at all: we can reject Schema 2 by config parameter and
> > +Docker Engine 1.10 will push the images in old format in accordance with the push strategy.
> > +
> > +## API changes
> > +
> > +* `MediaType` needs to be added to `ImageLayer`.
> > +* `Image.DockerImageMetadata` will be empty (except ID and Size). We don't have any data to fill it.
>
> We cannot drop this data - we are required to expose it. For schema 2, we
> should be including both the nested manifests AND the linked items to where
> we can read this info.
>
> Yes. I forgot about that. But if we use it, in fact we will still use the
> old manifest. If so, then migration to scheme 2 will be additional query
> for the old manifest.
> @smarterclayton <https://github.com/smarterclayton> We want to continue
> to use the scheme 1 ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8751/files/57fd3b1f09605bb86e13f58c41eb0e75058a67e5#r62209279>
>